orchard_commitments.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+import sys; assert sys.version_info[0] >= 3, "Python 3 required."
+
+from orchard_group_hash import group_hash
+from orchard_pallas import Fp, Scalar
+from orchard_sinsemilla import sinsemilla_hash_to_point
+from utils import i2lebsp
+
+# Commitment schemes used in Orchard https://zips.z.cash/protocol/nu5.pdf#concretecommit
+
+# https://zips.z.cash/protocol/nu5.pdf#constants
+L_ORCHARD_BASE = 255
+
+# https://zips.z.cash/protocol/nu5.pdf#concretehomomorphiccommit
+def homomorphic_pedersen_commitment(rcv: Scalar, D, v: Scalar):
+    return group_hash(D, b"v") * v + group_hash(D, b"r") * rcv
+
+def value_commit(rcv: Scalar, v: Scalar):
+    return homomorphic_pedersen_commitment(rcv, b"z.cash:Orchard-cv", v)
+
+def rcv_trapdoor(randbytes):
+    return Scalar.random(randbytes)
+
+# https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
+def sinsemilla_commit(r: Scalar, D, M):
+    assert isinstance(r, Scalar)
+    return sinsemilla_hash_to_point(D + b"-M", M).checked_incomplete_add(
+        group_hash(D + b"-r", b"") * r
+    )
+
+def sinsemilla_short_commit(r: Scalar, D, M):
+    return sinsemilla_commit(r, D, M).extract()
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteorchardnotecommit
+def note_commit(rcm, g_d, pk_d, v, rho, psi):
+    return sinsemilla_commit(
+        rcm,
+        b"z.cash: Orchard-NoteCommit",
+        g_d + pk_d + i2lebsp(64, v) + i2lebsp(L_ORCHARD_BASE, rho.s) + i2lebsp(L_ORCHARD_BASE, psi.s)
+    )
+
+def rcm_trapdoor(randbytes):
+    return Scalar.random(randbytes)
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteorchardnotecommit
+def commit_ivk(rivk: Scalar, ak: Fp, nk: Fp):
+    return Scalar(sinsemilla_short_commit(
+        rivk,
+        b"z.cash: Orchard-CommitIvk",
+        i2lebsp(L_ORCHARD_BASE, ak.s) + i2lebsp(L_ORCHARD_BASE, nk.s)
+    ).s)
```

Comment:
Hmm, in the spec, Commit<sup>ivk</sup>.Output is [defined as](https://zips.z.cash/protocol/nu5.pdf#abstractcommit) a scalar mod q<sub>P</sub>, not r<sub>P</sub>. And IVK parsing [MUST reject if the result is not in F<sub>q<sub>P</sub></sub>](https://zips.z.cash/protocol/nu5.pdf#orchardinviewingkeyencoding). So I think `commit_ivk` should not convert from a `Base` to a `Scalar` here:
```suggestion
    return sinsemilla_short_commit(
        rivk,
        b"z.cash: Orchard-CommitIvk",
        i2lebsp(L_ORCHARD_BASE, ak.s) + i2lebsp(L_ORCHARD_BASE, nk.s)
    )
```

The [Orchard key components spec](https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents) mentions that the output of Commit<sup>ivk</sup> is used as a KA<sup>Orchard</sup> private key `ivk`, so it's unclear there whether `ivk` is supposed to be typed as `Base` or `Scalar`, but given the above I think it should be typed `Base` and then explicitly converted to a `Scalar` at the point we use it. I'll make a similar change to the `orchard` crate.

cc @daira for verification.

Reply:
The protocol spec types ivk as an integer in [1, q_P).

Reply:
Yep, and that is enforced by construction: `SinsemillaCommit` never returns the identity (which we assert in the test vector generator for Sinsemilla), and thus the zero-output case of ExtractP is never triggered.

So my comment stands: change `ivk` to be typed as `Base` and then explicitly convert to `Scalar` when necessary.